src/connectors/impls/kafka.rs

```diff
@@ -76,3 +88,366 @@ fn is_failed_connect_error(err: &KafkaError) -> bool {
             | KafkaError::Global(RDKafkaErrorCode::UnknownTopicOrPartition | RDKafkaErrorCode::UnknownTopic | RDKafkaErrorCode::AllBrokersDown)
     )
 }
+
+struct TremorRDKafkaContext<Ctx>
+where
+    Ctx: Context + Send + Sync + 'static,
+{
+    ctx: Ctx,
+    connect_tx: Sender<KafkaError>,
+    metrics_tx: BroadcastSender<EventPayload>,
+    active: AtomicBool,
+}
+
+impl<Ctx> TremorRDKafkaContext<Ctx>
+where
+    Ctx: Context + Send + Sync + 'static,
+{
+    const PRODUCER: &'static str = "producer";
+    const CONSUMER: &'static str = "consumer";
+
+    fn new(
+        ctx: Ctx,
+        connect_tx: Sender<KafkaError>,
+        metrics_tx: BroadcastSender<EventPayload>,
+    ) -> Self {
+        Self {
+            ctx,
+            connect_tx,
+            metrics_tx,
+            active: AtomicBool::new(true),
+        }
+    }
+
+    fn on_connection_lost(&self) -> JoinHandle<()> {
+        let ctx = self.ctx.clone();
+
+        // only actually notify the notifier if we didn't do so before
+        // otherwise we would flood the connector and reconnect multiple times
+        if self.active.fetch_and(false, Ordering::AcqRel) {
```

Comment:
I think `swap` here would do the same thing as `fetch_and`, but be a bit more readable